### PR TITLE
[wip] Use libsodium for GetRandBytes

### PR DESCRIPTION
**Pull request by dagurval**
This is an alternative to https://github.com/bitcoin/bitcoin/pull/7891 used by, and cherry-picked from [zcash](https://github.com/zcash/zcash/issues/1632). Instead of implementing and maintaining strong randomness ourselves as Core has done (7891, 9821, 10338.. etc), [just use libsodium](https://paragonie.com/blog/2016/05/how-generate-secure-random-numbers-in-various-programming-languages).

Includes the following:
https://github.com/zcash/zcash/pull/793 - Add libsodium
https://github.com/zcash/zcash/pull/1268 - Upgrade libsodium for AVX2-detection bugfix.
https://github.com/zcash/zcash/pull/1706 - Use libsodium's CSPRNG instead of OpenSSL's
https://github.com/zcash/zcash/pull/2763/ - Update libsodium from 1.0.11 to 1.0.15

Tested on Linux and Mac.

### CODE REVIEW

File: depends/packages/libsodium.mk [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+$(package)_download_path=https://download.libsodium.org/libsodium/releases/
+$(package)_file_name=libsodium-1.0.8.tar.gz
+$(package)_sha256_hash=c0f191d2527852641e0a996b7b106d2e04cbc76ea50731b2d0babd3409301926
+$(package)_dependencies=
+$(package)_config_opts=
+
+define $(package)_preprocess_cmds
+  cd $($(package)_build_subdir); ./autogen.sh
+endef
+
```
Comment (dgenr8) [inline depends/packages/libsodium.mk]: Should we add
```
define $(package)_set_vars
  $(package)_config_opts_linux=--with-pic
endef
```

**Comment by dgenr8** [comment]
I pushed tag `sodium` to test gitian build.  It's failing with the following in var/build.log:
```
make[4]: Entering directory `/home/ubuntu/build/bitcoin/depends/work/build/i686-pc-linux-gnu/libsodium/1.0.15-2da31405804/src/libsodium'
  CC       crypto_shorthash/siphash24/libsodium_la-shorthash_siphashx24.lo
  CC       crypto_shorthash/siphash24/ref/libsodium_la-shorthash_siphashx24_ref.lo
gcc: error: 1.0.15": No such file or directory
gcc: error: 1.0.15": No such file or directory
```

**Comment by dagurval** [comment]
Looking at verbose gcc output, the error looks like an extra `"` on `DPACKAGE_VERSION`. 

```
libtool: compile:  gcc -m32 -std=gnu99 -DPACKAGE_NAME=\"libsodium\" -DPACKAGE_TARNAME=\"libsodium\" -DPACKAGE_VERSION=\"1.0.16\" "-DPACKAGE_STRING=\"libsodium 1.0.16\"" -DPACKAGE_BUGREPORT=\"https://github.com/jedisct1/libsodium/issues\" -DPACKAGE_URL=\"https://github.com/jedisct1/libsodium\" -DPACKAGE=\"libsodium\" -DVERSION=\"1.0.16\" -DHAVE_PTHREAD_PRIO_INHERIT=1 -DHAVE_PTHREAD=1 -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_C_VARARRAYS=1 -DHAVE_CATCHABLE_SEGV=1 -DHAVE_CATCHABLE_ABRT=1 -DTLS=__thread -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_MMINTRIN_H=1 -DHAVE_EMMINTRIN_H=1 -DHAVE_PMMINTRIN_H=1 -DHAVE_TMMINTRIN_H=1 -DHAVE_SMMINTRIN_H=1 -DHAVE_AVXINTRIN_H=1 -DHAVE_AVX2INTRIN_H=1 -DHAVE_WMMINTRIN_H=1 -DHAVE_RDRAND=1 -DHAVE_SYS_MMAN_H=1 -DNATIVE_LITTLE_ENDIAN=1 -DHAVE_CPUID=1 -DASM_HIDE_SYMBOL=.hidden -DHAVE_WEAK_SYMBOLS=1 -DCPU_UNALIGNED_ACCESS=1 -DHAVE_ATOMIC_OPS=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DHAVE_MMAP=1 -DHAVE_MLOCK=1 -DHAVE_MADVISE=1 -DHAVE_MPROTECT=1 -DHAVE_NANOSLEEP=1 -DHAVE_POSIX_MEMALIGN=1 -DHAVE_GETPID=1 -DCONFIGURED=1 -I. -I./include/sodium -I./include/sodium -msse2 -mssse3 -msse4.1 -mavx -mavx2 -I/home/ubuntu/build/bitcoin/depends/i686-pc-linux-gnu/include -pipe -O2 -pthread -fvisibility=hidden -fPIC -fno-strict-aliasing -fno-strict-overflow -fstack-protector -c crypto_stream/salsa20/xmm6int/salsa20_xmm6int-avx2.c  -fPIC -DPIC -o crypto_stream/salsa20/xmm6int/libavx2_la-salsa20_xmm6int-avx2.o
gcc: error: 1.0.16": No such file or directory
```

**Comment by dagurval** [comment]
I don't feel like debugging libsodium's build system on a trusty image — inside Gitian. We can see if the problem persists after upgrading Gitian to use bionic (wip at Core PR#13171). Zcash may not have encountered this issue because their Gitian image is based on debian/trusty.

**Pull request closed without merging**